file/file.go

```diff
@@ -9,18 +9,35 @@ import (
 )
 
 // SafeWrite writes data to a temporary file and then renames it to what's specified in path
-func SafeWrite(path string, data []byte) error {
+func SafeWrite(path string, data []byte) (err error) {
 	dir := filepath.Dir(path)
-	err := os.MkdirAll(dir, 0755)
+
+	err = os.MkdirAll(dir, 0755)
 	if err != nil {
-		return err
+		return
 	}
 
-	tmpPath := path + ".tmp"
+	// Ensure multiple simultaneous callers will not choose the same file.
+	tmpFile, err := ioutil.TempFile(dir,  "tmp")
+	if err != nil {
+		return
+	}
+
+	tmpPath := tmpFile.Name()
+	// Don't leak temp files left by failed write attempts
+	defer func() {
+		tmpFile.Close()
+		if err != nil {
+			os.Remove(tmpPath)
+		}
+	}()
+
 	err = ioutil.WriteFile(tmpPath, data, 0644)
```

Comment:
I don't like the idea of opening the temporary file with ioutil.TempFile() and *at the same time* using ioutil.WriteFile() to write to this file, because ioutil.WriteFile() will open the second file descriptor.
Please consider using tmpFile.Write() instead.

Reply:
done